### PR TITLE
first version of dicom support histoqc

### PR DESCRIPTION
**Pull request by nanli-emory**
#252  - Refactor BaseImage Class to support reading multiple WSI format such DICOM By using Abstract Base Class

**Add:**
1. histoqc/wsihandles/**WSIImageHandle.py** - wsi image handle (ABC) 
2. histoqc/wsihandles/**DicomHandle.py** - wsi dicom support handle
3. histoqc/wsihandles/**OpenSlideHandle.py** - openslide support handle

**Modify:**
1. histoqc/**BaseImage.py** - refactor to support and merge **WSIImageHandle**, **DicomHandle** and **OpenSlideHandle**
2. histoqc/**BasicModule.py** - change how to get the basic stats info method.
3. histoqc/**ClassificationModule.py** - fixed #254 - correct Classification Module
4. histoqc/**SaveModule.py** - change for getting macro and label image
5. histoqc/**TileExtractionModule.py** - change slide image handle from openslide to WSIImageHandle

### CODE REVIEW

File: histoqc/BaseImage.py
Comment (choosehappy) [inline histoqc/BaseImage.py]: should import dicomhandle and openslide handle be in this file? i would have thought they would have only been needed in the WSIImageHandle class?

File: histoqc/BaseImage.py
Comment (choosehappy) [inline histoqc/BaseImage.py]: maybe this goes in the WSI handle file?

as well, I'm wondering if these should be an if statement or a try/catch statement?

or maybe a try catch should be in place around the if? 

i guess in this case its obvious which file type to use? might need to check with petros

File: histoqc/ClassificationModule.py
Comment (choosehappy) [inline histoqc/ClassificationModule.py]: i'm not entirely sure i understand this code here. if you have say 4 classes, won't think always produce a binary mask based on the 1st channel because of line 210? 

```
mask = cal[:, :, 1] > thresh
```

is this generalizable to multiple classes, and/or what is the situation where we have more than 2 classes? how does the user specify that which channel they want to be the defining class? 

it seems like we may only want to support 2 channel masks to keep life simple, and rely on the user to provide the correct type of input?

File: histoqc/SaveModule.py
Comment (choosehappy) [inline histoqc/SaveModule.py:75]: should this be refactored into a single function along with he saveMacro function? the code looks identical except for the string variable?

File: histoqc/wsihandles/OpenSlideHandle.py
Comment (choosehappy) [inline histoqc/wsihandles/OpenSlideHandle.py:19]: is it common to start private methods with two underscores instead of one? I've not seen this style before

File: histoqc/wsihandles/__init__.py
Comment (choosehappy) [inline histoqc/wsihandles/__init__.py]: need to be careful with this approach as it creates a tight coupling to a new dependency, wsidicom

if wsidicom is not installed, even if the user doesn't try to use this on a DICOM image, it will still crash i think?

there are a few things I'm thinking:

1) what is the dependency size of WSIdicom? for example, if it is 10GB of installation that is one thing, if it is 100k that is another, the latter of which justifies adding in wsidicom into the requirements.txt

2) if it is a larger install, we may not want to assume the user has it, and thus only users which enable dicom support should have access to it. this can be done via e.g., the config file receiving a list of the all the handlers to use when reading WSI, this would thus extend to other reading libraries, like bioformats

3) i think one way of doing this is that there is a list created of each of the WSI handlers, in the order specified by the user, for example, in the config file

-config 
wsi_handlers= "openslidehandler","dicom","type3","type4..." 

-code
for wsi_handler in wsi_handlers:
          osh = wsi_handler.open(fname)
          if osh
                     return osh
if osh == None:
          error, no handler could read this image



notice that this logic is a bit more clean, because each handler gets to see the image (in the order specified by the user) and gets to choose if it wants to load the image or not, so it is the handler programmer that writes the code. in the version here, you see that the same concept code ("is directory" function) is one level higher, making it the main branch developers responsibility to integrate as opposed to the handler developers  job.

the interesting point is that in this manner, one need only import the libraries that the user specifies, breaking the tight coupling of having to install everything. 

for our purposes this is likely fine its only 2 handlers, and the dependencies are minimal, but for a larger scale project with more developers, or more handlers, the way above is definitely more correct. 

do you see the difference?  perhaps you want to take a stab at refactoring that way for your own understanding, i think it shouldb't be more than a couple minutes of effort

**Review by choosehappy — CHANGES_REQUESTED** [review]

File: histoqc/BaseImage.py
Comment (nanli-emory) [reply histoqc/BaseImage.py]: I can't import DicomHandle and OpenSlideHandle in WSIImageHandle class since DicomHandle and OpenSlideHandle need to import WSIImageHandle as the superclass. It causes an error if I import DicomHandle and OpenSlideHandle in WSIImageHandle class:
```ImportError: cannot import name 'WSIImageHandle' from partially initialized module 'histoqc.wsihandles.WSIImageHandle' (most likely due to a circular import)```

File: histoqc/BaseImage.py
Comment (nanli-emory) [reply histoqc/BaseImage.py]: I still think about this part. We probably should use factory pattern + dynamic import to implement how to create a subclass for ABC class. 
Do you have any thoughts on this?  @CielAl

File: histoqc/SaveModule.py
Comment (nanli-emory) [reply histoqc/SaveModule.py:75]: Actually, we have a single function named **saveAssociatedImage** with the string parameter. The **saveLabel** and **saveMacro** functions are restrictions of using **saveAssociatedImage** since wsidicom and openslide image and APIs are heterogeneous. In openslide, **associated_images** property is _**Images, such as label or macro images, which are associated with this slide. This is a Mapping from the image name to RGBA Image.**_ We can directly get the associated image by the key such as **openslide.associated_images["macro"]**. However, in wsidicom, there is no associated image property. The wsidicom only provides 2 functions to get associated images - **WsiDicom.read_label()** and **WsiDicom.read_overview()**. 
Saving label and macro functions should be the most frequently used functions. This is why I create these 2 functions individually. Open to any better proposal. ;)

File: histoqc/wsihandles/OpenSlideHandle.py
Comment (nanli-emory) [reply histoqc/wsihandles/OpenSlideHandle.py:19]: Yes. Please see the following test code for the private functions.
```python
class Student:
    def _not_private(self):  # not private function
	    print('This is not private function.')
                 
    def __private(self): # private function
	    print('This is private function.')

std = Student()
# call no private function
std._not_private()

# Using name mangling to call private function
std._Student__private()

# AttributeError
std.__private() 
```
But some people said _Python's **private** functions are not actually private since user can use name mangling to call private function_
reference: https://www.geeksforgeeks.org/private-methods-in-python/

File: histoqc/BaseImage.py
Comment (CielAl) [reply histoqc/BaseImage.py]: @nanli-emory @choosehappy 
I agree.
If I understand correctly about what you want here:
(1) avoid importing each subclass handles because not all might be installed on the client's machine.
(2) dynamically create the corresponding subclass handle based on what's installed or any input arguments.

So herein you may have the below options:
(1) create a factory class (or make it extending Protocol for Python 3.10 and above) that implements the relevant APIs and the builder:
`
    
class HandleBuilder:

    @staticmethod
    def build(params) -> WSIImageHandle:

        ... # return different handle types based on 

`

(2) Optionally, rather than just making a HandleBuilder, make the supposed factory builder a higher level of abstraction in some way close to the skeletal pattern. Herein your WSIImageHandle would be the abstract class to manage how to open the image resources and compute corresponding attributes by the data read accordingly.
You can create another interface (e.g., `class ImageCore` )that 
    --- (1)  associate a WSIImageHandle as a private attribute.
    --- (2) provide all APIs of WSIImageHandle by returning the results of the associated WSIImageHandle 
    --- (3) implement the factory builder function above as a static method or class method

By doing this the advantage is, you can move certain behavior like "get_best_level_for_downsample" from BaseImage to `ImageCore` and therefore your BaseImage can focus on carrying data/mask/results between each module. 

(3) Simply use a function as the factory builder.
image

File: histoqc/ClassificationModule.py
Comment (nanli-emory) [reply histoqc/ClassificationModule.py]: Hi @choosehappy,
It is relevant to #250 and #254. My explanation is on #254.

File: histoqc/BaseImage.py
Comment (CielAl) [reply histoqc/BaseImage.py]: @nanli-emory @choosehappy 
Note that essentially you want to hide the details of all WSIImageHandle's subclasses from BaseImage as HistoQC's pipeline itself shouldn't care about it anyway, therefore the import of all subclasses should be avoided in the BaseImage class. 

The purpose of the second proposal is essentially to organize the BaseImage class so that all WSI-related methods that should not be in a dict subclass (BaseImage) can be moved to this layer. I am personally kinda wary of extending a python built-in type and lump too many overrides there. 

In some way, the BaseImage at the moment is deeply coupled to all modules and most of the modules access BaseImage by string keys -- note IDEs usually have limited functionality to refactor string literals, so even if you fully annotated these keys by Literal, it will be annoying to manage future changes.

With this abstraction, you can fold the dict of BaseImage as an associated attribute, refactor the corresponding helper function of compression, and integrate all corresponding key-value pairs as the unified interface:
e.g. you can make the BaseImage the abstraction I mentioned but associate the actual dict as an attribute besides the WSIImageHandle,
`
class BaseImage:


    __base_image_dict: dict
    __wsi_image_handle:

 
    # build from dict of params
    def __init__(self, params):
         self.__wsi_image_handle = some-factory-builder
         self.__base_image_dict = another-builder-for-init-baseimagedict-from-params


`
and therefore with IDE it's easier to refactor the Modules for future changes and can make fewer mistakes (e.g., typos of keys).

File: histoqc/ClassificationModule.py
Comment (choosehappy) [reply histoqc/ClassificationModule.py]: left a comment there. i still don't understand why this is needed or why we're not always doing binary classification? shouldn't need any multi-class stuff

File: histoqc/SaveModule.py
Comment (choosehappy) [reply histoqc/SaveModule.py:75]: i think you can remove both of these functions, and have them be keywords used directly in "saveAssociatedImage" via the config file?

you basically already have this, except instead of having "key" come in as a method-level parameter, you read it from the config file?

File: histoqc/wsihandles/OpenSlideHandle.py
Comment (choosehappy) [reply histoqc/wsihandles/OpenSlideHandle.py:19]: super cool, i didn't know about this functionality, thanks!

File: histoqc/wsihandles/__init__.py
Comment (choosehappy) [reply histoqc/wsihandles/__init__.py]: did you see this comment as well?

File: histoqc/BaseImage.py
Comment (choosehappy) [reply histoqc/BaseImage.py]: "Note that essentially you want to hide the details of all WSIImageHandle's subclasses from BaseImage as HistoQC's pipeline itself shouldn't care about it anyway, therefore the import of all subclasses should be avoided in the BaseImage class."

Yes, exactly this. this is what it looks like, which also addresses the first comment in this review and last comment:

```python
#---main.py
from baseclass import AbstracthbaseClass

user_list=["openslide","dicom"]

for types in user_list:
    ci = AbstracthbaseClass.returnClass(types)
    print(ci)
    print(ci.has_bounding_box)
from abc import ABC, abstractmethod

#---baseclass.py
class AbstracthbaseClass(ABC):


    @staticmethod
    def returnClass(type):
        if type=="openslide":
            print("c1")
            from c1class import c1
            return c1(type)
        else:
            print("c2")
            from c2class import c2
            return c2(type)

    @property
    @abstractmethod
    def has_bounding_box(self):
        ...

#--- c1class.py
from baseclass import AbstracthbaseClass

class c1(AbstracthbaseClass):
    def __init__(self, fname):
        self.fname = fname
        print("c1 init")

    @property
    def has_bounding_box(self):
        print("c1 bounding box")
        return self.fnamefrom baseclass import AbstracthbaseClass

#----c2class.py
class c2(AbstracthbaseClass):
    def __init__(self, fname):
        self.fname = fname
        print("c2 init")

    @property
    def has_bounding_box(self):
        print("c2 bounding box")

```

Comment (nanli-emory) [reply histoqc/BaseImage.py]: Hi @choosehappy, I Changed how to load and create subclasses for WSIImageHandle. But do we really need to have "handles" parameters for ImageBase? 
Currently, **create_wsi_handle** function will auto-determine the file type and import the subclass if needed.

File: histoqc/wsihandles/WSIImageHandle.py
Comment (choosehappy) [inline histoqc/wsihandles/WSIImageHandle.py]: where did you get this list of file extensions? just want to make sure its exhaustive

File: histoqc/wsihandles/WSIImageHandle.py
Comment (choosehappy) [inline histoqc/wsihandles/WSIImageHandle.py]: i've not seen this before, do you have a reference you can link to or something similar? trying to understand how robust this may be

File: histoqc/wsihandles/__init__.py
Comment (choosehappy) [reply histoqc/wsihandles/__init__.py]: @nanli-emory , please see this comment and confirm that you saw it?

File: histoqc/wsihandles/WSIImageHandle.py
Comment (choosehappy) [inline histoqc/wsihandles/WSIImageHandle.py]: i think you may be trying too hard here, you're trying to come up with a narrow definition of what a dicom is, and I'm not sure that that is going to work, e.g., i know that all of them don't end in dcm, and it is highly dependent upon the manufacturer

as well, the user should know about their environment

i would suggest modifying this to be a try-catch situation based on the readers the user has in their config. i feel like i made this comment before, did you have some concern about it? i just pointed it out again, can you please confirm that you saw it?

lets think about this a little deeper. what do we ultimately want? the handler to try to and open the file, right?
what is wrong with something like this:

```python
for reader in config_reader_list
	import reader
	try:
		osh = reader.open(fname)
	exception:
		pass
if osh == None:
	error -- no reader wants to read this file
	
```

where now the *reader itself* gets to determine if it wants to open the file, as opposed to your approach here which requires both a sophisticated algorithm to see if it is a dicom, and as well a list of file extensions

ultimately, if the reader can open it, its a success, so we can use that directly as a success criteria?

since the user is providing the list of readers that are set up in their system, we need not worry about importing things which they don't think are enabled

thoughts?

**Review by choosehappy — CHANGES_REQUESTED** [review]
Getting there! requesting better refactoring to simplify reader selection

File: histoqc/wsihandles/WSIImageHandle.py
Comment (nanli-emory) [reply histoqc/wsihandles/WSIImageHandle.py]: Hi @choosehappy I got it from [openslide website](https://github.com/choosehappy/HistoQC/pull/256).

File: histoqc/wsihandles/WSIImageHandle.py
Comment (nanli-emory) [reply histoqc/wsihandles/WSIImageHandle.py]: Hi @choosehappy, I saw this on pydicom code. https://github.com/pydicom/pydicom/blob/d21ecad8e663c8937d064da561bf7299415d7d98/pydicom/misc.py#L39

File: histoqc/wsihandles/WSIImageHandle.py
Comment (nanli-emory) [reply histoqc/wsihandles/WSIImageHandle.py]: Hi @choosehappy  I move the create_wsi_handle into WSIImageHandle.

File: histoqc/wsihandles/WSIImageHandle.py
Comment (choosehappy) [reply histoqc/wsihandles/WSIImageHandle.py]: got it, thanks. i still think we should move away from this approach as discussed in my github review, did you see it?

File: histoqc/wsihandles/WSIImageHandle.py
Comment (choosehappy) [reply histoqc/wsihandles/WSIImageHandle.py]: is this comment intended in response to what i said above?

File: histoqc/wsihandles/WSIImageHandle.py
Comment (CielAl) [reply histoqc/wsihandles/WSIImageHandle.py]: It's probably not ideal to identify which reader to use by extension, especially when what these readers can handle are not mutually exclusive.

You might want to define a default option if nothing is specified or whatever library requested by the user is not installed, e.g., import these readers by a specified order and use whatever is correctly imported. Otherwise, it should be defined by the user, e.g., put a string or int placeholder in the config file to specify what type of WSIHandle it is.

**Comment by nanli-emory** [comment]
hi @choosehappy,
can I merge this PR if there are no more code changes as required?

**Comment by choosehappy** [comment]
lets wait for some feedback from petros first

File: requirements.txt
Comment (petroslk) [inline requirements.txt:13]: bumping up to Pillow~=9.1.0 should solve dependency conflicts with wsidicom

**Pull request merged**
